roles/metrics-export.lua

```diff
@@ -1,17 +1,274 @@
+                host = "0.0.0.0"
+            end
+        end
+    elseif type(listen) == "number" then
+        host = "0.0.0.0"
+        port = listen
+    end
+
+    if type(port) == "number" and (port < 1 or port > 65535) then
+        return nil, nil, "port must be in the range [1, 65535]"
+    end
+    return host, port, nil
+end
+
+local http_handlers = {
+    json = function(req)
+        local json_exporter = require('metrics.plugins.json')
+        return req:render({ text = json_exporter.export() })
+    end,
+    prometheus = function(...)
+        local http_handler = require('metrics.plugins.prometheus').collect_http
+        return http_handler(...)
+    end,
+}
+-- It is used as an error string with the predefined order.
+local http_supported_formats_str = "json, prometheus"
+
+local function validate_http_endpoint(endpoint)
+    if type(endpoint) ~= "table" then
+        error("http endpoint must be a table, got " .. type(endpoint), 4)
+    end
+    if endpoint.path == nil then
+        error("http endpoint 'path' must exist", 4)
+    end
+    if type(endpoint.path) ~= "string" then
+        error("http endpoint 'path' must be a string, got " .. type(endpoint.path), 4)
+    end
+    if string.sub(endpoint.path, 1, 1) ~= '/' then
+        error("http endpoint 'path' must start with '/', got " .. endpoint.path, 4)
+    end
+
+    if endpoint.format == nil then
+        error("http endpoint 'format' must exist", 4)
+    end
+    if type(endpoint.format) ~= "string" then
+        error("http endpoint 'format' must be a string, got " .. type(endpoint.format), 4)
+    end
+
+    if not http_handlers[endpoint.format] then
+        error("http endpoint 'format' must be one of: " ..
+              http_supported_formats_str .. ", got " .. endpoint.format, 4)
+    end
+end
+
+local function validate_http_node(node)
+    if type(node) ~= "table" then
+        error("http configuration node must be a table, got " .. type(node), 3)
+    end
+
+    local _, _, err = parse_listen(node.listen)
+    if err ~= nil then
+        error("failed to parse http 'listen' param: " .. err, 3)
+    end
+
+    node.endpoints = node.endpoints or {}
+    if type(node.endpoints) ~= "table" then
+        error("http 'endpoints' must be a table, got " .. type(node.endpoints), 3)
+    end
+    if not is_array(node.endpoints) then
+        error("http 'endpoints' must be an array, not a map", 3)
+    end
+    for _, endpoint in ipairs(node.endpoints) do
+        validate_http_endpoint(endpoint)
+    end
+
+    for i, ei in ipairs(node.endpoints) do
+        local pathi = remove_side_slashes(ei.path)
+        for j, ej in ipairs(node.endpoints) do
+            if i ~= j then
+                local pathj = remove_side_slashes(ej.path)
+                if pathi == pathj then
+                    error("http 'endpoints' must have different paths", 3)
+                end
+            end
+        end
+    end
 end
 
-local function apply()
+local httpd_servers = nil
```

Comment:
I may be wrong since there aren't a lot of enthusiasts (and I'm not one of them), but such approach seems to not support hot reload.

What does I mean when I'm talking about hot reload?
- instance is running and working;
- replace the module files;
- run `package.loaded[*] = nil`;
- next time the `require`s are called they will fetch new files.

In case one will do it with this implementation, the following is likely to happen.
- old servers are still somewhere in the memory and will be cleaned up by garbage collector later (it will likely close everything);
- new servers will be started after `apply()`;

which is likely to result in port clash.

I don't known whether code hot reload is supported when working with Tarantool 3 with configuration at all, though. Your case is likely to be fixed if you will use some named `_G` variable instead of file local variable. (But some new questions will arise, like "what to do if I'm required for the first time and there are already some servers".) You can also see [crud stash](https://github.com/tarantool/crud/blob/master/crud/common/stash.lua) which do the same with cartridge compatibility (we don't need one here, but maybe we'll need to add something similar if something similar to cartridge roles hot reload will be introduced). Or you may just write a small note in README that code is not hot-reload ready. Or just ignore it.

Reply:
> but such approach seems to not support hot reload

The Tarantool 3 have not the hotreload by design. The approach is fine and used in the `config` module everywhere, see:

https://github.com/tarantool/tarantool/blob/b0b32bff022caeffd64ca687baab908f25fa2fcb/src/box/lua/config/applier/sharding.lua#L6-L7

Reply:
Well, there should be no issues with two watchers doing the same idempotent thing, but it won't work for two servers binding on the same port.

Reply:
A role will not re-require until unload:

https://github.com/tarantool/tarantool/blob/b0b32bff022caeffd64ca687baab908f25fa2fcb/src/box/lua/config/applier/roles.lua#L3-L4